<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:
- **TASK 1:** Mark all launch sites on a map
- **TASK 2:** Mark the success/failed launches for each site on the map
- **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [9]:
!pip3 install folium
!pip3 install wget
!pip3 install pandas

In [10]:
import folium
import wget
import pandas as pd

In [11]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site. 


In [17]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)

100% [................................................................................] 7710 / 7710

Now, you can take a look at what are the coordinates for each site.


In [19]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [22]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example, 


In [24]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle. 


Now, let's add a circle for each launch site in data frame `launch_sites`


_TODO:_  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [32]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
for i in range (len(launch_sites_df.index)):
    coordinate = [launch_sites_df["Lat"][i], launch_sites_df["Long"][i]]
    circle = folium.Circle(coordinate, radius=100, color='#d35400', fill=True).add_child(folium.Popup(launch_sites_df["Launch Site"][i]))
    marker = folium.map.Marker(
        coordinate,
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % launch_sites_df["Launch Site"][i],
            )
        )
    site_map.add_child(circle)
    site_map.add_child(marker)

site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:
- Are all launch sites in proximity to the Equator line?
- Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


### Analysis of the Exercise and Explanation of Findings

The objective of this exercise is to visually map the locations of SpaceX launch sites using the `folium` library to interpret the geographical latitude and longitude coordinates effectively. Here is a detailed analysis and explanation of the findings.

#### 1. Data Preparation and Analysis
- **Dataset:** The `spacex_launch_geo.csv` contains the columns `Launch Site`, `Lat`, `Long`, and `class`.
- **Relevant Columns Selected:** We focus on the columns `Launch Site`, `Lat`, and `Long` to identify the locations of the launch sites.
- **List of Launch Sites:**
  - **CCAFS LC-40**: 28.562302°N, -80.577356°W
  - **CCAFS SLC-40**: 28.563197°N, -80.576820°W
  - **KSC LC-39A**: 28.573255°N, -80.646895°W
  - **VAFB SLC-4E**: 34.632834°N, -120.610745°W

#### 2. Map Visualization
The `folium` library was used to plot the sites:
- **Initial Map Center:** NASA Johnson Space Center in Houston, Texas.
- **Circles and Markers:** Added circles with a 100-meter radius and customized markers for each launch site with popup labels displaying the launch site names.

#### 3. Visual Findings
Upon generating and exploring the map, the following observations can be made:

1. **Proximity to the Equator:**
   - **Observation:** None of the launch sites are particularly close to the Equator. The majority of SpaceX sites in Florida (CCAFS LC-40, CCAFS SLC-40, and KSC LC-39A) are situated around latitude 28°N. The VAFB SLC-4E site in California is even further north, at latitude 34°N.
   - **Explanation:** Although not near the Equator, these latitudes are sufficiently low to offer certain orbital advantages for launches to equatorial or inclined orbits.

2. **Proximity to the Coast:**
   - **Observation:** All launch sites are located very close to the coast.
   - **Explanation:** Coastal proximity is essential for space launches due to several reasons:
     - **Safety:** In the event of launch failures, rockets can fall into the ocean, minimizing risk to the population.
     - **Safe Launch Direction:** Launching eastward over the ocean takes advantage of Earth’s rotation and ensures safety.

### Summary of Findings:
- **Latitudinal Location:** The sites are positioned at mid-latitudes, which are suitable for SpaceX's launch needs.
- **Coastal Proximity:** This is a key factor in site selection, ensuring safety and optimizing launch trajectories.

These points provide insight into why SpaceX and other space companies choose specific sites for launches: a balance of safety, efficiency, and logistical convenience for achieving various orbital goals.


# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [39]:
spacex_df.tail(10)

Launch Site        Lat       Long  class
46    KSC LC-39A  28.573255 -80.646895      1
47    KSC LC-39A  28.573255 -80.646895      1
48    KSC LC-39A  28.573255 -80.646895      1
49  CCAFS SLC-40  28.563197 -80.576820      1
50  CCAFS SLC-40  28.563197 -80.576820      1
51  CCAFS SLC-40  28.563197 -80.576820      0
52  CCAFS SLC-40  28.563197 -80.576820      0
53  CCAFS SLC-40  28.563197 -80.576820      0
54  CCAFS SLC-40  28.563197 -80.576820      1
55  CCAFS SLC-40  28.563197 -80.576820      0

Next, let's create markers for all launch records. 
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [43]:
marker_cluster = MarkerCluster()


_TODO:_ Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [45]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red


In [46]:
# Function to assign color to launch outcome
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'
    
spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
spacex_df.tail(10)

Launch Site        Lat       Long  class marker_color
46    KSC LC-39A  28.573255 -80.646895      1        green
47    KSC LC-39A  28.573255 -80.646895      1        green
48    KSC LC-39A  28.573255 -80.646895      1        green
49  CCAFS SLC-40  28.563197 -80.576820      1        green
50  CCAFS SLC-40  28.563197 -80.576820      1        green
51  CCAFS SLC-40  28.563197 -80.576820      0          red
52  CCAFS SLC-40  28.563197 -80.576820      0          red
53  CCAFS SLC-40  28.563197 -80.576820      0          red
54  CCAFS SLC-40  28.563197 -80.576820      1        green
55  CCAFS SLC-40  28.563197 -80.576820      0          red

_TODO:_ For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [48]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, 
# e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for index, record in spacex_df.iterrows():
    # TODO: Create and add a Marker cluster to the site map
    # marker = folium.Marker(...)
    marker = folium.Marker(
        location=[record['Lat'], record['Long']],  # Latitude and Longitude
        icon=folium.Icon(color='white', icon_color=record['marker_color'])  # Set the color based on outcome
    )
    marker_cluster.add_child(marker)

site_map

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [56]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [59]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

_TODO:_ Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [61]:
# find coordinate of the closet coastline
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
# Example coordinates of launch site and coastline
launch_site_lat = 28.56341
launch_site_lon = -80.57679
coastline_lat = 28.56367
coastline_lon = -80.56772

# Calculate the distance to the coastline
distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
print(f"Distance to coastline: {distance_coastline:.2f} km")


Distance to coastline: 0.89 km


_TODO:_ After obtained its coordinate, create a `folium.Marker` to show the distance


In [63]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
distance_marker = folium.Marker(
    location=[coastline_lat, coastline_lon],
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html=f'<div style="font-size: 12; color:#d35400;"><b>{distance_coastline:.2f} KM</b></div>'  # Formatting distance
           )
)
site_map.add_child(distance_marker)

_TODO:_ Draw a `PolyLine` between a launch site to the selected coastline point


In [65]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
# Create a PolyLine object to connect the launch site and coastline
coordinates = [[launch_site_lat, launch_site_lon], [coastline_lat, coastline_lon]]
lines = folium.PolyLine(locations=coordinates, weight=1, color='blue')

# Add the PolyLine to the map
site_map.add_child(lines)

# Display the map
site_map

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


_TODO:_ Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


# Another Example of How to Do It with Added Icons


In [76]:
# Coordinates for the launch site and points of interest
launch_site_lat = 28.56341
launch_site_lon = -80.57679
highway_lat = 28.56357
highway_lon = -80.57081
city_lat = 28.07923
city_lon = -80.6051
railway_lat = 28.57221
railway_lon = -80.58528
coastline_lat = 28.56367
coastline_lon = -80.56772

# Function to calculate the distance between two geographical points
def calculate_distance(lat1, lon1, lat2, lon2):
    R = 6373.0  # Approximate Earth radius in km
    lat1, lon1, lat2, lon2 = map(radians, [lat1, lon1, lat2, lon2])
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))
    distance = R * c
    return distance

# Create a map centered on the launch site
site_map = folium.Map(location=[launch_site_lat, launch_site_lon], zoom_start=15)

# List of points of interest with names, colors, and icons
points_of_interest = [
    {"name": "Highway", "lat": highway_lat, "lon": highway_lon, "color": "blue", "icon": "road"},
    {"name": "City", "lat": city_lat, "lon": city_lon, "color": "red", "icon": "building"},
    {"name": "Railway", "lat": railway_lat, "lon": railway_lon, "color": "purple", "icon": "train"},
    {"name": "Coastline", "lat": coastline_lat, "lon": coastline_lon, "color": "cadetblue", "icon": "water"}
]

# Add markers and distance lines for each point of interest
for poi in points_of_interest:
    distance = calculate_distance(launch_site_lat, launch_site_lon, poi['lat'], poi['lon'])
    
    # Create a marker at the point of interest with an icon and color
    marker = folium.Marker(
        location=[poi['lat'], poi['lon']],
        popup=f"{poi['name']} - {distance:.2f} KM",
        icon=folium.Icon(color=poi['color'], icon=poi['icon'], prefix='fa')
    )
    site_map.add_child(marker)

    # Draw a line between the launch site and the point of interest
    line = folium.PolyLine(
        locations=[[launch_site_lat, launch_site_lon], [poi['lat'], poi['lon']]],
        weight=2,
        color=poi['color']
    )
    site_map.add_child(line)

# Add a marker for the launch site
launch_marker = folium.Marker(
    location=[launch_site_lat, launch_site_lon],
    popup="Launch Site",
    icon=folium.Icon(color='green', icon='rocket', prefix='fa')
)
site_map.add_child(launch_marker)

# Display the map
site_map


After you plot distance lines to the proximities, you can answer the following questions easily:
- Are launch sites in close proximity to railways?
- Are launch sites in close proximity to highways?
- Are launch sites in close proximity to coastline?
- Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


# Analysis of Launch Site Proximity to Key Infrastructure

After plotting the distance lines between the launch site and the points of interest (railways, highways, coastline, and cities), we can answer the following questions based on the calculated distances. Here's a breakdown:

## 1. Are launch sites in close proximity to railways?

- **Distance to Railway**: The calculated distance between the launch site at `28.56341, -80.57679` and the railway at `28.57221, -80.58528` is approximately **1.28 km**.

**Answer**: Yes, the launch site is in close proximity to the railway. A distance of around 1.28 km suggests that railways are quite nearby, which could be an important consideration for transporting rocket parts.

## 2. Are launch sites in close proximity to highways?

- **Distance to Highway**: The distance to the highway at `28.56357, -80.57081` is calculated to be around **0.89 km** from the launch site.

**Answer**: Yes, the launch site is very close to a highway. At just under 1 km, this proximity is practical for transporting equipment and personnel, making highways a key infrastructure component for launch site accessibility.

## 3. Are launch sites in close proximity to coastline?

- **Distance to Coastline**: The coastline at `28.56367, -80.56772` is approximately **0.89 km** from the launch site.

**Answer**: Yes, the launch site is in close proximity to the coastline. The 0.89 km distance is relatively small, and this is likely because launch sites near coastlines are strategically chosen to avoid risks and to provide ample space for rocket launches over water, reducing risks to nearby populated areas.

## 4. Do launch sites keep a certain distance away from cities?

- **Distance to City**: The calculated distance to the city at `28.07923, -80.6051` is **~5.88 km**.

**Answer**: Yes, the launch site does maintain a certain distance away from cities. A distance of 5.88 km is quite significant, indicating that launch sites are generally located far from populated urban areas to minimize risks associated with rocket launches, such as accidents or noise disturbance.

---

## Summary of Findings:

- **Proximity to Railways**: The launch site is quite close to the railway (1.28 km), which is useful for transportation of large rocket parts.
- **Proximity to Highways**: The launch site is very close to a highway (0.89 km), which provides easy access for vehicles.
- **Proximity to Coastline**: The launch site is in close proximity to the coastline (0.89 km), ensuring safe launch trajectories over the ocean, far from populated areas.
- **Distance from Cities**: The launch site maintains a reasonable distance from the city (5.88 km), which is likely done for safety reasons and to avoid impacting urban areas.

---

## Explanation:

Launch sites are typically located near railways and highways to facilitate the transportation of equipment and materials. Proximity to a coastline is essential for safety, as it ensures the launch is over water, reducing the risk of damage or injury to populated areas in the event of an accident. The distance from cities is a standard safety measure to avoid any harm to people or infrastructure. 

These observations reflect how the site’s location is carefully chosen, balancing accessibility, safety, and minimizing potential risks.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/)


### Other Contributors


Joseph Santarcangelo


## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2021-05-26|1.0|Yan|Created the initial version|


**⚠️ Note:** These exercises were solved by **Crisel Nublo** 🪻. 

Copyright © 2021 IBM Corporation. All rights reserved.
